ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerSelectionActions.hs

```diff
@@ -119,34 +136,42 @@ withPeerSelectionActions
         localRootsVar)
       (\thread -> k thread peerSelectionActions)
   where
-    -- We first try to get public root peers from the ledger, but if it fails
-    -- (for example because the node hasn't synced far enough) we fall back
-    -- to using the manually configured bootstrap root peers.
-    requestPublicRootPeers :: Int -> m (Map peeraddr (PeerAdvertise, IsLedgerPeer), DiffTime)
-    requestPublicRootPeers n = do
-      peers_m <- getLedgerPeers (NumberOfPeers $ fromIntegral n) AllLedgerPeers
-      case peers_m of
-           -- No peers from Ledger
-           Nothing    -> do
-             (m, dt) <- requestConfiguredRootPeers n
-             let m' = Map.map (\a -> (a, IsNotLedgerPeer)) m
-             return (m', dt)
+    -- We start by reading the current ledger state judgement, if it is
+    -- YoungEnough we only care about fetching for ledger peers, otherwise we
+    -- aim to fetch bootstrap peers.
+    requestPublicRootPeers :: LedgerPeersKind -> Int -> m (PublicRootPeers peeraddr, DiffTime)
+    requestPublicRootPeers ledgerPeersKind n = do
+      -- Check if the node is in a sensitive state
+      isSensitive <- atomically $ isInSensitiveState <$> readUseBootstrapPeers
+                                                     <*> readLedgerStateJudgement
+      case isSensitive of
+        -- If the ledger state is not in a sensitive state we should get ledger
+        -- peers, the Nothing case should not happen but there can be a race
+        -- condition. If that's the case we try again soon enough.
+        False -> do
+          mbLedgerPeers <- getLedgerPeers (NumberOfPeers $ fromIntegral n) ledgerPeersKind
+          case mbLedgerPeers of
+            -- no peers from the ledger
+            Nothing -> do
+              (publicRootPeers, dt) <- requestConfiguredPublicRootPeers n
+              pure (PublicRootPeers.fromMapAndSet publicRootPeers Set.empty Set.empty Set.empty, dt)
```

Comment:
We should have functions:

* `PublicRootPeers.fromPublicRootPeersMap`,
* `PublicRootPeers.fromLedgerPeers`,
* `PublicRootPeers.fromBootstrapPeers`, and
* `PublicRootPeers.fromBigLedgerPeers`

instead of `PublicRootPeers.fromMapAndSet`.